### ISSUE TITLE
F523 No longer works if *args is present

### ISSUE THREAD

**Issue by colin99d** [charliermarsh/ruff#2567]
The following code triggers F523:
`"{0}".format(1, 2)`

However, the code below does not:
`"{0}".format(1, 2, *args)`

This seems incorrect to me, because both function perform the same operation:
```
>>> args = [1, 2, 3]
>>> "{0}".format(1, 2)
'1'
>>> "{0}".format(1, 2, *args)
'1'
```

**Comment by dhruvmanila** [comment]
When a starred argument is present, it's difficult to determine whether the number of arguments is equal to the number of placeholders. The reason being that `*args` can expand into unknown number of elements at runtime and we cannot know it statically.

One solution could be to parallelly process both the placeholder and format argument one after the other and stop at starred argument or the last placeholder, whichever comes first. If we encounter the starred argument first, then we know that at least all the previous positional arguments are correct. On the other hand, if there are no placeholders remaining but we have _positional arguments_, those will be reported. Note that here the starred argument will also be reported.

***Example:***
```python
"{0} {1}".format(a, *args)  # All good here

"{0} {1}".format(a, b, *args)
#                      ^^^^^

"{0} {1}".format(a, b, c, *args)
#                      ^  ^^^^^
```

**Comment by charliermarsh** [comment]
Yeah, I _think_ we can only flag when the number of non-starred position arguments is greater than the number of placeholders.

The debatable case is `"{0} {1}".format(a, b, *args)`... Here, this _could_ work if `*args` is an empty tuple, I think. But it's obviously a bit strange.

**Comment by dhruvmanila** [comment]
> The debatable case is `"{0} {1}".format(a, b, *args)`... Here, this _could_ work if `*args` is an empty tuple, I think. But it's obviously a bit strange.

I'm not sure how is this debatable as we've got 2 placeholders where `a` and `b` will go and thus `*args` is unused.

There's a special case where we should ignore - when there's only a single starred argument - but I suppose this would satisfy the logic where `{0}` is mapped to `*args` and this rule is not about _unused placeholders_.
```python
"{0} {1}".format(*args)
```

**Comment by charliermarsh** [comment]
Sorry, you're right. What I was trying to point out was that `args` _can_ be an empty tuple, but it'd be strange not to flag in that case.

**Comment by dhruvmanila** [comment]
Yes, I do think that we should be flagging it. The variable is being unused which is what the rule is about. The contents of the variable doesn't matter as it's not used in the first place.

**Comment by charliermarsh** [comment]
Agreed.

**Comment by dhruvmanila** [comment]
Well, Python raises an error if the `*args` unpacks to nothing. So, then should we skip checking the `*args` argument and instead just check all the other positional arguments irrespective of whether `*args` is present or not.

```python
args = []

"{0}{1}".format(1, *args)
# ---------------------------------------------------------------------------
# IndexError                                Traceback (most recent call last)
# Cell In [2], line 1
# ----> 1 "{0}{1}".format(1, *args)
#
# IndexError: Replacement index 1 out of range for positional args tuple

args.append(2)

"{0}{1}".format(1, *args)
# '12'
```

**Comment by dhruvmanila** [comment]
> _**Example:**_
> 
> ```python
> "{0} {1}".format(a, *args)  # All good here
> 
> "{0} {1}".format(a, b, *args)
> #                      ^^^^^
> 
> "{0} {1}".format(a, b, c, *args)
> #                      ^  ^^^^^
> ```

`mypy` only triggers for the third example, so I guess it's skips checking the `*args` argument.

> mypy: Not all arguments converted during string formatting
